index/test/rtree/rtree_with_strategies.cpp

```diff
@@ -0,0 +1,156 @@
+    rtree.query(bgi::within(QueryValue{}), std::back_inserter(values));
+}
+
+template
+<
+    typename QueryValue, typename Rtree,
+    std::enable_if_t<(!is_bp_or_bb_v<QueryValue, Rtree>), int> = 0
+>
+void test_queries_bp_bb(Rtree const& ) {}
+
+template
+<
+    typename QueryValue, typename Rtree,
+    std::enable_if_t<is_pb_or_bb_v<QueryValue, Rtree>, int> = 0
+>
+void test_queries_pb_bb(Rtree const& rtree)
+{
+    // These predicates use algorithms that are not implemented for
+    // some geometry combinations
+    std::vector<typename Rtree::value_type> values;
+    rtree.query(bgi::contains(QueryValue{}), std::back_inserter(values));
+    rtree.query(bgi::covers(QueryValue{}), std::back_inserter(values));
+}
+
+template
+<
+    typename QueryValue, typename Rtree,
+    std::enable_if_t<(!is_pb_or_bb_v<QueryValue, Rtree>), int> = 0
+>
+void test_queries_pb_bb(Rtree const& ) {}
+
+template
+<
+    typename QueryValue, typename Rtree,
+    std::enable_if_t<is_pp_or_bb_v<QueryValue, Rtree>, int> = 0
+>
+void test_overlaps_pp_bb(Rtree const& rtree)
+{
+    std::vector<typename Rtree::value_type> values;
+    rtree.query(bgi::overlaps(QueryValue{}), std::back_inserter(values));
+    rtree.query(bgi::touches(QueryValue{}), std::back_inserter(values));
+}
+
+template
+<
+    typename QueryValue, typename Rtree,
+    std::enable_if_t<(!is_pp_or_bb_v<QueryValue, Rtree>), int> = 0
+>
+void test_overlaps_pp_bb(Rtree const& ) {}
+
+template <typename QueryValue, typename Rtree>
+void test_queries(Rtree const& rtree)
+{
+    std::vector<typename Rtree::value_type> values;
+    rtree.query(bgi::intersects(QueryValue{}), std::back_inserter(values));
+    rtree.query(bgi::nearest(QueryValue{}, 1), std::back_inserter(values));
+    test_queries_bp_bb<QueryValue>(rtree);
+    test_queries_pb_bb<QueryValue>(rtree);
+    test_overlaps_pp_bb<QueryValue>(rtree);
+}
+
+template <typename Value, typename Params, typename Strategies>
+void test_strategies()
+{
+    std::vector<Value> values;
+    bgi::rtree<Value, bgi::parameters<Params, Strategies>> rtree{values};
+    rtree.insert(Value{});
+    rtree.remove(Value{});
+
+    test_queries<point>(rtree);
+    test_queries<box>(rtree);
+    test_queries<segment>(rtree);
+}
+
+template <typename Value, typename Params>
+void test_params()
+{
+    test_strategies<Value, Params, bg::strategies::index::cartesian<>>();
+    test_strategies<Value, Params, bg::strategies::cartesian<>>();
+    test_strategies<Value, Params, bg::strategies::index::spherical<>>();
+    test_strategies<Value, Params, bg::strategies::spherical<>>();
+    test_strategies<Value, Params, bg::strategies::index::geographic<>>();
+    test_strategies<Value, Params, bg::strategies::geographic<>>();
+}
+
+template <typename Value>
+void test_value()
+{
+    test_params<Value, bgi::linear<4>>();
+    test_params<Value, bgi::quadratic<4>>();
+    test_params<Value, bgi::rstar<4>>();
```

Comment:
Did we document somewhere what is the adviced option? `rstar` ?

Reply:
It depends on the data and algorithm using the rtree. It's described here: https://www.boost.org/doc/libs/1_87_0/libs/geometry/doc/html/geometry/spatial_indexes/introduction.html#geometry.spatial_indexes.introduction.r_tree

With that said I'd probably use quadratic balancing algorithm by default and packing if possible.